content/posts/creating-grafana-dashboards-for-nodejs-on-kubernetes/index.md

```diff
@@ -0,0 +1,248 @@
+
+### The Counter Metric Type
+
+In this graph, we used the counter metric type, which is continuously increasing. Most of the time we're interested in the the relative increase in a range. The `rate()` function calculates this and takes into account the exact time between each two data points. It also accounts for counter resets, which may happen when an instance restarts.
+
+### The Gauge Metric Type
+
+Unlike counters, gauges are metrics that can also go down. As such they are very intuitive to use and do not require derivation or the use of the `rate()` function.
+Examples for gauges could be temperature or CPU load. However, measuring CPU load using gauges may not be advised, since the gauge only represents a snapshot at sampling time and does not take into account any variation between two samples. Therefore tools like Prometheus Node Exporter export CPU usage in seconds as a counter.
+
+## Request Performance
+
+Our application's metrics endpoint exports `http_request_duration_seconds_bucket` metrics with different values `n` for the `le` label, which indicates, how many requests where shorter than `n` seconds. These "buckets" together form a histogram metric.
+
+One caveat is, that the buckets in this case are not exlusive. A request that took 0.7 seconds is counted in the buckets `le: 1.5`, `le: 10` and `le: +Inf` once each. This simplifies calculating ratios and percentages, but requires a subtraction to get information about how many requests were e.g. between 1.5 and 10 seconds.
+
+This histogram allows us to monitor whether if our application performs well and within service level objectives. We'll look into different ways of visualizing the histogram data. Choose the one that suits you and your use case best.
+
+### Histogram Graph Charts
+
+The most simple chart just displays all the buckets as graph charts individually
+
+![Graph chart showing all request buckets stacked](./request-duration-stacked_all.png)
+
+The following query calculates the rate for each bucket and pod and then sums that up by bucket to produce the chart above.
+
+```promql
+sum by (le) (rate(http_request_duration_seconds_bucket{kubernetes_namespace="awesome-node"}[3m]))
+```
+
+A drawback of this chart is, that most of the requests are in the bucket `le: 0.003` and the graphs representing other buckets all get crammed into a very tiny space.
+
+We can exclude the bucket for the shortest requests and remove them from the sum by turning the buckets around and calculating the request counts that are larger than the `le` thresholds.
+
+![Graph chart showing request buckets excluding the fastest bucket](./request-duration-stacked_reduced.png)
+
+Use this query to produce this chart:
+
+```promql
+sum by (app, le) (rate(http_request_duration_seconds_count{app="awesome-node"}[3m]) - ignoring (le) group_right rate(http_request_duration_seconds_bucket{le!="+Inf"}[3m]))
```

Comment:
I'm adding the following text:

This query uses a one-to-many matching to subtract many metrics (the buckets) from a single metric (the count). In order to support this, we need to tell Prometheus that the right part has the higher cardinality using the group_right modifier.

Additionally, Prometheus by default matches metrics whose labels match the same. However, the le label is only available on the buckets. For the matching operation to succeed, we tell it to ignore the le label, by adding the ignoring (le) modifier.

Reply:
Cool! That would do the trick I think!

One correction in the new text. I would suggest

`Furthermore, by default Prometheus matches metrics with matching label values. However, in this case the le label ...`

instead of

`Additionally, Prometheus by default matches metrics whose labels match the same. However, the le label ...`